iothub/device/src/Message.cs

```diff
@@ -51,28 +59,27 @@ public Message(Stream stream)
         }
 
         /// <summary>
-        /// Constructor which uses the input byte array as the body
+        /// Constructor which uses the input byte array as the body.
         /// </summary>
-        /// <param name="byteArray">a byte array which will be used to
-        /// form the body stream</param>
-        /// <remarks>user should treat the input byte array as immutable when
-        /// sending the message.</remarks>
-        public Message(
-            byte[] byteArray)
-            : this(new MemoryStream(byteArray))
+        /// <remarks>User should treat the input byte array as immutable when sending the message.</remarks>
+        /// <param name="byteArray">A byte array which will be used to form the body stream.</param>
+        /// <param name="setDefaultMessageId">A flag indicating if the MessageId should be set to a random GUID.</param>
+        public Message(byte[] byteArray, bool setDefaultMessageId = false)
+            : this(new MemoryStream(byteArray), setDefaultMessageId)
         {
-            // reset the owning of the steams
+            // reset the owning of the streams
             _ownsBodyStream = true;
         }
 
         /// <summary>
         /// This constructor is only used on the Gateway http path so that
         /// we can clean up the stream.
         /// </summary>
-        /// <param name="stream"></param>
-        /// <param name="ownStream"></param>
-        internal Message(Stream stream, bool ownStream)
-            : this(stream)
+        /// <param name="stream">A stream which will be used as body stream.</param>
+        /// <param name="ownStream">A flag indicating that the client library will dispose the stream when the Message is disposed.</param>
+        /// <param name="setDefaultMessageId">A flag indicating if the MessageId should be set to a random GUID.</param>
+        internal Message(Stream stream, bool ownStream, bool setDefaultMessageId = false)
```

Comment:
This can potentially cause confusion with
```csharp
public Message(Stream stream, bool setDefaultMessageId = false)
```
 but this method is internal; so we can be careful about its usage.
This does come across to me as bad code though.

Reply:
Yeah, I don't love that.

Reply:
I've used named args wherever `internal Message(Stream stream, bool ownStream, bool setDefaultMessageId = false)` is currently used, but I don't see a way around this.

Reply:
Updated to use an enum